src/core/lib/support/orphanable.h

```diff
@@ -0,0 +1,166 @@
+/*
+ *
+ * Copyright 2017 gRPC authors.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ *
+ */
+
+#ifndef GRPC_CORE_LIB_SUPPORT_ORPHANABLE_H
+#define GRPC_CORE_LIB_SUPPORT_ORPHANABLE_H
+
+#include <grpc/support/log.h>
+#include <grpc/support/sync.h>
+
+#include <memory>
+
+#include "src/core/lib/debug/trace.h"
+#include "src/core/lib/support/abstract.h"
+#include "src/core/lib/support/debug_location.h"
+#include "src/core/lib/support/memory.h"
+
+namespace grpc_core {
+
+// A base class for orphanable objects.
+class Orphanable {
+ public:
+  // Gives up ownership of the object.  The implementation must arrange
+  // to destroy the object without further interaction from the caller.
+  virtual void Orphan() GRPC_ABSTRACT;
+
+  // Not copyable or movable.
+  Orphanable(const Orphanable&) = delete;
+  Orphanable& operator=(const Orphanable&) = delete;
+
+  GRPC_ABSTRACT_BASE_CLASS
+
+ protected:
+  Orphanable() {}
+  virtual ~Orphanable() {}
+};
+
+template <typename T>
+class OrphanableDelete {
+ public:
+  void operator()(T* p) { p->Orphan(); }
+};
+
+template <typename T, typename Deleter = OrphanableDelete<T>>
+using OrphanablePtr = std::unique_ptr<T, Deleter>;
```

Comment:
according to https://github.com/grpc/proposal/blob/master/L6-allow-c%2B%2B-in-grpc-core.md we shouldn't be using `std::unique_ptr` directly.

Reply:
The intent of that is that code outside of lib/support will use the abstractions provided by lib/support instead of directly using things provided in the standard library.  But code in lib/support can use things provided by the standard library as long as we don't introduce a run-time dependency.  In the case of `std::unique_ptr<>`, the only part that would impose a run-time dependency is the deleter, which we're overriding here.

Note that we do already define `grpc_core::UniquePtr<>`, which is basically defined as `std::unique_ptr<>` with a deleter that calls `gpr_free()`:

https://github.com/grpc/grpc/blob/master/src/core/lib/support/memory.h#L51